### DISCUSSION TITLE
How to define `globalThis.__DEV__` in compile time?

### DISCUSSION THREAD

**Discussion by lionskape** [vercel/next.js#65355] [Help]
### Summary

Is there any way to extend compile time replacement definitions?
Apollo Client uses `globalThis.__DEV__` to [handle production/development](https://www.apollographql.com/docs/react/development-testing/reducing-bundle-size#why-use-globalthis__dev__-instead-of-processenvnode_env) builds.

So I have to add `globalThis.__DEV__` for Apollo, there is some ways:

1. I can add it to webpack config, but I will lose a turbopack feature:
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack(config, { webpack }) {
    config.plugins.push(
      new webpack.DefinePlugin({
        "globalThis.__DEV__": false,
      })
    );
    return config;
  },
};

module.exports = nextConfig;
```

2. Maybe there is a way to extend swc config in next? I do not found a place, where I can put something like this in nextjs project
```
{
  "jsc": {
    "transform": {
      "optimizer": {
        "globals": {
          "vars": {
            "globalThis.__DEV__": "false"
          }
        }
      }
    }
  }
}
```

**Comment by maximpostnikov** [comment]
This happens because of double serialization (JSON.stringify) and overly strict typing (Record<string, string>).
As a result, 'false' becomes "false" (a truthy string) instead of false.

For now, you can work around it by bypassing the type:

```
const nextConfig: NextConfig = {
  compiler: {
    define: {
      'globalThis.__DEV__': false as never
    }
  }
}
```

This works for me on Next.js 16.2.1

**Reply by Anwesha-Bali** [reply]
Thanks, this helped!

**Comment by phryneas** [comment]
I don't know why this old topic is being picked up again, but for everyone asking themselves this question, and before more LLM-generated answers chime in:

It's 2026, update to Apollo Client 4. 
`globalThis.__DEV__` is only a fallback for very outdated bundlers there, with a modern bundler it will already bundle correctly through the `development` and `production` exports conditions, nothing to set up.

**Reply by snyamathi** [reply]
@phryneas https://www.apollographql.com/docs/react/development-testing/reducing-bundle-size#nextjs

I don't know if the Apollo documentation is outdated, but they seem to still recommend doing this step.

**Reply by phryneas** [reply]
Yeah, we need to get that updated. I'm an Apollo Client maintainer, so take my word for it here :)

If your bundler uses a `production` export condition and you use Apollo Client 4.x, you won't need `globalThis.__DEV__` for Apollo Client, as no code you bundle will have references to that.
That said, you might still need it for the upstream `graphql` package.

**Comment by selcukcukur** [comment]
@lionskape Right now Next.js doesn’t expose a direct way to extend SWC’s `jsc.transform.optimizer.globals` in your project config, so you can’t drop in that JSON block the way you showed. That’s why most people fall back to `webpack.DefinePlugin` — but as you noticed, that disables Turbopack.  

The practical options today are :

- **Use `webpack.DefinePlugin`**: works, but you lose Turbopack.  
- **Set it at runtime**: instead of compile‑time replacement, you can define `globalThis.__DEV__` early in your app (e.g. in `next.config.js` via `env`, or in a small bootstrap file before Apollo is imported). For example:
  ```js
  if (typeof globalThis.__DEV__ === 'undefined') {
    globalThis.__DEV__ = process.env.NODE_ENV !== 'production'
  }
  ```
  This avoids compile‑time replacement but still gives Apollo what it expects.  
- **Wait for Turbopack support**: the Next.js team has been working on exposing more SWC optimizer options. At the moment, there isn’t a documented way to extend those globals in `next.config.js`.  

So the short answer: there’s no official hook for extending SWC compile‑time definitions in Next.js yet. If you want true compile‑time replacement, you’ll need to stick with `webpack.DefinePlugin`. If you want to keep Turbopack, the workaround is to set `globalThis.__DEV__` at runtime before Apollo loads.  

That way you don’t get 100 calls in draft mode, and you don’t lose Turbopack features.

**Comment by sumitkumar2374** [comment]
1. Vite
In your vite.config.js:

```JavaScript:

export default {
  define: {
    'globalThis.__DEV__': process.env.NODE_ENV !== 'production',
  },
};
```

2. Webpack
Use the DefinePlugin in your webpack.config.js:

```JavaScript:

const webpack = require('webpack');

module.exports = {
  plugins: [
    new webpack.DefinePlugin({
      'globalThis.__DEV__': JSON.stringify(process.env.NODE_ENV !== 'production'),
    }),
  ],
}; 
```

3. Babel
If you want to handle it during transpilation, use babel-plugin-transform-define:

```JSON:

{
  "plugins": [
    ["transform-define", {
      "globalThis.__DEV__": "process.env.NODE_ENV !== 'production'"
    }]
  ]
}
```
